src/PostgREST/AppState.hs

```diff
@@ -0,0 +1,110 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module PostgREST.AppState
+  ( AppState
+  , getConfig
+  , getDbStructure
+  , getIsWorkerOn
+  , getPgVersion
+  , getTime
+  , init
+  , initWithPool
+  , killMainThread
+  , pool
+  , putConfig
+  , putDbStructure
+  , putIsWorkerOn
+  , putPgVersion
+  , releasePool
+  ) where
+
+import qualified Hasql.Pool as P
+
+import Control.AutoUpdate (defaultUpdateSettings, mkAutoUpdate,
+                           updateAction)
+import Data.IORef         (IORef, atomicWriteIORef, newIORef,
+                           readIORef)
+import Data.Time.Clock    (UTCTime, getCurrentTime)
+
+import PostgREST.Config                (AppConfig (..))
+import PostgREST.DbStructure           (DbStructure)
+import PostgREST.DbStructure.PgVersion (PgVersion (..))
+
+import Protolude      hiding (toS)
+import Protolude.Conv (toS)
+
+
+data AppState = AppState
+  { statePool         :: P.Pool -- | Connection pool, either a 'Connection' or a 'ConnectionError'
+  -- | Used to sync the listener(NOTIFY reload) with the connectionWorker. No
+  -- connection for the listener at first. Only used if dbChannelEnabled=true.
+  , statePgVersion    :: MVar PgVersion
+  -- | No schema cache at the start. Will be filled in by the connectionWorker
+  , stateDbStructure  :: IORef (Maybe DbStructure)
+  -- | Helper ref to make sure just one connectionWorker can run at a time
+  , stateIsWorkerOn   :: IORef Bool
+  -- | Config that can change at runtime
+  , stateConf         :: IORef AppConfig
+  , stateGetTime      :: IO UTCTime
+  , stateMainThreadId :: ThreadId
+  }
+
+init :: AppConfig -> IO AppState
+init conf = do
+  newPool <- initPool conf
+  initWithPool newPool conf
+
+initWithPool :: P.Pool -> AppConfig -> IO AppState
+initWithPool newPool conf =
+  AppState newPool
+    <$> newEmptyMVar
+    <*> newIORef Nothing
+    <*> newIORef False
+    <*> newIORef conf
+    <*> mkAutoUpdate defaultUpdateSettings { updateAction = getCurrentTime }
+    <*> myThreadId
+
+initPool :: AppConfig -> IO P.Pool
+initPool AppConfig{..} =
+  P.acquire
+    ( configDbPoolSize
+    , fromRational . toRational $ configDbPoolTimeout
```

Comment:
I don't know why this happens. The following change:

```patch
index 3b88022..0b293e5 100644
--- a/src/PostgREST/Config.hs
+++ b/src/PostgREST/Config.hs
@@ -46,6 +46,7 @@ import Data.List               (lookup)
 import Data.List.NonEmpty      (fromList, toList)
 import Data.Maybe              (fromJust)
 import Data.Scientific         (floatingOrInteger)
+import Data.Time.Clock         (NominalDiffTime)
 import Numeric                 (readOct, showOct)
 import System.Environment      (getEnvironment)
 import System.Posix.Types      (FileMode)
@@ -66,7 +67,7 @@ data AppConfig = AppConfig
   , configDbExtraSearchPath     :: [Text]
   , configDbMaxRows             :: Maybe Integer
   , configDbPoolSize            :: Int
-  , configDbPoolTimeout         :: Int
+  , configDbPoolTimeout         :: NominalDiffTime
   , configDbPreRequest          :: Maybe Text
   , configDbPreparedStatements  :: Bool
   , configDbRootSpec            :: Maybe Text
@@ -195,7 +196,7 @@ parser optPath env dbSettings =
     <*> optWithAlias (optInt "db-max-rows")
                      (optInt "max-rows")
     <*> (fromMaybe 10 <$> optInt "db-pool")
-    <*> (fromMaybe 10 <$> optInt "db-pool-timeout")
+    <*> (fromIntegral <$> parsePoolTimeout)
     <*> optWithAlias (optString "db-pre-request")
                      (optString "pre-request")
     <*> (fromMaybe True <$> optBool "db-prepared-statements")
@@ -224,6 +225,9 @@ parser optPath env dbSettings =
     <*> (fmap T.unpack <$> optString "server-unix-socket")
     <*> parseSocketFileMode "server-unix-socket-mode"
   where
+    parsePoolTimeout :: C.Parser C.Config Int
+    parsePoolTimeout = fromMaybe 10 <$> optInt "db-pool-timeout"
+
     parseAppSettings :: C.Key -> C.Parser C.Config [(Text, Text)]
     parseAppSettings key = addFromEnv . fmap (fmap coerceText) <$> C.subassocs key C.value
       where
```

Raise the following error in io tests:

```
postgrest: FatalError {fatalErrorMessage = "Error in config ParseError \"/run/user/1001/pytest-of-remo/pytest-30/test_stable_config_sigusr2_set1/config:7:21:\\n  |\\n7 | db-pool-timeout = 10s\\n  |                     ^\\nunexpected 's'\\nexpecting '.', 'E', 'e', digit, end of input, or end of line\\n\""}
```

This change should not change anything about how "10s" is parsed, no?

Reply:
Since this is the "stable config" test, I would assume this is about `--dump-config`. You didn't adjust the `show configDbPoolTimeout`. I would think you need to convert it to integer there first, because we can't parse the `10s` that are `show`n from the `NominalDiffTime` type.

Reply:
Ah, it was failing on the reload! Thank you @wolfgangwalther ! Fixed in latest commit

> Edit: Ah, really cool to see the "stable config test" providing value here :)

Yes, awesome!